pandas/core/arrays/interval.py

```diff
@@ -0,0 +1,838 @@
+    If None, dtype will be inferred
+
+    .. versoinadded:: 0.23.0
+
+Attributes
+----------
+left
+right
+closed
+mid
+length
+values
+is_non_overlapping_monotonic
+
+Methods
+-------
+from_arrays
+from_tuples
+from_breaks
+%(extra_methods)s\
+
+%(examples)s\
+
+Notes
+------
+See the `user guide
+<http://pandas.pydata.org/pandas-docs/stable/advanced.html#intervalindex>`_
+for more.
+
+See Also
+--------
+Index : The base pandas Index type
+Interval : A bounded slice-like interval; the elements of an IntervalIndex
+interval_range : Function to create a fixed frequency IntervalIndex
+cut, qcut : Convert arrays of continuous data into Categoricals/Series of
+            Intervals
+"""
+
+
+@Appender(_interval_shared_docs['class'] % dict(
+    klass="IntervalArray",
+    summary="Pandas array for interval data that are closed on the same side",
+    versionadded="0.23.0",
+    name='', extra_methods='', examples='',
+))
+class IntervalArray(IntervalMixin, ExtensionArray):
+    dtype = IntervalDtype()
+    ndim = 1
+    can_hold_na = True
+    _na_value = _fill_value = np.nan
+
+    def __new__(cls, data, closed=None, dtype=None, copy=False,
+                fastpath=False, verify_integrity=True):
+
+        from pandas.core.indexes.interval import IntervalIndex
+
+        if fastpath:
+            return cls._simple_new(data.left, data.right, closed,
+                                   copy=copy, dtype=dtype,
+                                   verify_integrity=False)
+
+        if isinstance(data, ABCSeries) and is_interval_dtype(data):
+            data = data.values
+        if isinstance(data, (cls, IntervalIndex)):
+            left = data.left
+            right = data.right
+            closed = data.closed
+        else:
+
+            # don't allow scalars
+            if is_scalar(data):
+                msg = ("{}(...) must be called with a collection of some kind,"
+                       " {} was passed")
+                raise TypeError(msg.format(cls.__name__, data))
+
+            data = maybe_convert_platform_interval(data)
+            left, right, infer_closed = intervals_to_interval_bounds(data)
+
+            if _all_not_none(closed, infer_closed) and closed != infer_closed:
+                # GH 18421
+                msg = ("conflicting values for closed: constructor got "
+                       "'{closed}', inferred from data '{infer_closed}'"
+                       .format(closed=closed, infer_closed=infer_closed))
+                raise ValueError(msg)
+
+            closed = closed or infer_closed
+
+        return cls._simple_new(left, right, closed, copy=copy, dtype=dtype,
+                               verify_integrity=verify_integrity)
+
+    @classmethod
+    def _simple_new(cls, left, right, closed=None,
+                    copy=False, dtype=None, verify_integrity=True):
+        result = IntervalMixin.__new__(cls)
+
+        closed = closed or 'right'
+        left = _ensure_index(left, copy=copy)
```

Comment:
this is SO duplicating II
have to do something about this - this is a technical debt nightmare

Reply:
Did you see IntervalIndex._simple_new? There isn't any duplicated code between the two.

Reply:
so EA takes a different take on construction that in Index. Where we use _simple_new and _shalllow_copy, but conceptually these are the same things. Would be ok with changing the spellings inside Index to conform this. This is my main objection generally. We spell things one way and a different way in EA, this is so confusing.

Reply:
@jreback Can you elaborate on this comment (as you say this is your main objection, but I don't fully understand it).

So `IntervalArray._simple_new` construct from left/right, while `IntervalIndex._simple_new` now constructs from values (IntervalArray), while previously it constructed from left/right. So yes, this is different between array and index, but, for IntervalIndex it is now more consistent with the base Index._simple_new (which constructs from values), so I think that is a good thing.